stl/inc/flat_set

```diff
@@ -435,12 +445,22 @@ private:
         const iterator _End   = end();
         const iterator _Where = lower_bound(_Val);
         if constexpr (_Multi) {
+            _STL_INTERNAL_STATIC_ASSERT(is_same_v<remove_cvref_t<_Ty>, _Kty>);
             return _Cont.emplace(_Where, _STD forward<_Ty>(_Val));
         } else {
             if (_Where != _End && _Keys_equal(*_Where, _Val)) {
                 return pair{_Where, false};
             }
-            return pair{_Cont.emplace(_Where, _STD forward<_Ty>(_Val)), true};
+            if constexpr (is_same_v<remove_cvref_t<_Ty>, _Kty>) {
+                return pair{_Cont.emplace(_Where, _STD forward<_Ty>(_Val)), true};
+            } else {
+                // flat_set::insert(auto&&)
+                _STL_INTERNAL_STATIC_ASSERT(_Keylt_transparent && is_constructible_v<_Kty, _Ty>);
+                _Kty _Keyval{_STD forward<_Ty>(_Val)};
+                _STL_ASSERT(lower_bound(_Keyval) == _Where && !_Keys_equal(_Keyval, *_Where),
+                    "find(val) was not equal to find(key_type{forward(val)})");
+                return pair{_Cont.emplace(_Where, _STD move(_Keyval)), true};
+            }

```

Comment:
This is for debug of `flat_set::insert([hint,]auto&&)`; there are fairly dangerous use patterns like:
```cpp
    std::flat_set<int, std::less<>> s{ 1,2,3,5,7 };
    s.insert(2.5);
```
Which can break invariant unexpectedly. See details in [this comment](https://github.com/microsoft/STL/pull/3985#issuecomment-1691614165).

Otherwise, this is actually equivalent to the `if constexpr (is_same_v<remove_cvref_t<_Ty>, _Kty>)` branch (as `_Cont.emplace(_Where, _STD forward<_Ty>(_Val))` can does the same thing).

**Due to the trouble it introduces, I believe `flat_set::insert([hint,]auto&&)` should be renamed in the standard to avoid accidental misuse.**

Reply:
I don't get why this is dangerous. Wouldn't `2.5` be converted to `2` first and cause the insertion to fail? I tested the 2.5 insertion with `flat_map` and the flat_map was unchanged.

Reply:
find(2.5) whould get `end()`; find(int(2.5))->find(2) would get 2; this is precondition violation and is UB.

Reply:
For `flat_set::emplace`, I think a conversion to the `key_type` is required before insertion, as this is explicitly required in `flat_multiset::emplace`:
![image](https://github.com/microsoft/STL/assets/60953653/aecec8bb-72a6-45b0-862c-9bac51ee47e4)
Conversion happens before insertion even if `flat_multiset` can have transparent comparators...

![image](https://github.com/microsoft/STL/assets/60953653/b1f2507b-df26-4ace-a859-bb517bd4b128)
This means I cannot call `emplace(_hint)` directly, as in `emplace`, for non-key types I need to firstly convert `auto&&` to a `key_type`, which may change the value of `auto&&`.

Reply:
> find(2.5) whould get `end()`; find(int(2.5))->find(2) would get 2; this is precondition violation and is UB.
> 
> See [this link](https://github.com/microsoft/STL/pull/3985#issuecomment-1691024595)

So the implementation is not wrong then ... ? It's just counter-intuitive?

Reply:
I'm still not confident whether the implementation is correct for `insert([hint,]auto&&)`; but yes, `s.insert(2.5);` is just UB behavior, and I think this method is too tricky and can cause a lot of surprises in practice.